ci/run-tests.sh

```diff
@@ -1,10 +1,38 @@
-#!/bin/bash
-
-cargo test \
-    --features wasi-threads \
-    --features wasi-http \
-    --features component-model \
-    --features serve \
-    --workspace \
-    --exclude test-programs \
-    $@
+#!/usr/bin/env bash
+
+FEATURES="wasi-threads,wasi-http,component-model,serve,"
+
+
+while true; do
+    stderr=$(mktemp)
+
+    echo "info: running tests with cargo features: $FEATURES"
+    set -x
+    cargo test \
+          --workspace \
+          --features "$FEATURES" \
+          --exclude test-programs \
+          $@ \
+          2> >(tee -a "$stderr" >&2)
+    code=$?
+    set +x
+
+    case "$code" in
+        "0")
+            exit 0
+            ;;
+        *)
+            if grep -q 'error: none of the selected packages contains these features: ' "$stderr"; then
+                feature_to_remove=$(
+                    sed 's/error: none of the selected packages contains these features: //g' "$stderr" \
+                        | sed -e 's/,.*//g'
+                )
+                echo "info: disabling cargo feature '$feature_to_remove' and retrying..."
+                FEATURES=$(echo "$FEATURES" | sed -e "s/$feature_to_remove,//g")
+                continue
+            fi
+
+            exit "$code"
+            ;;
+    esac
+done
```

Comment:
I think it might be best to simplify this and replace it with `--all-features`. Most of the above features are on-by-default but it's also not an exhaustive list of off-by-default features that should be tested for crates. For example `wasmtime-environ` should have the `compile` feature enabled when testing, which it is today due to feature unioning but if it's split to its own job then it's probably best to slap `--all-features` and call it a day here.